src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -0,0 +1,630 @@
+            _gnuUnusedBytes ??= new byte[FieldLengths.AllGnuUnused];
+
+            long actualLength = GetTotalDataBytesToWrite();
+            TarEntryType actualEntryType = GetCorrectTypeFlagForFormat(TarFormat.Gnu);
+
+            int checksum = WriteName(buffer, out _);
+            checksum += WriteCommonFields(buffer, actualLength, actualEntryType);
+            checksum += WriteGnuMagicAndVersion(buffer);
+            checksum += WritePosixAndGnuSharedFields(buffer);
+            checksum += WriteGnuFields(buffer);
+            WriteChecksum(checksum, buffer);
+
+            archiveStream.Write(buffer);
+
+            if (_dataStream != null)
+            {
+                WriteData(archiveStream, _dataStream, actualLength);
+            }
+        }
+
+        // Writes the current header as a PAX Extended Attributes entry into the archive stream.
+        private void WriteAsPaxExtendedAttributes(Stream archiveStream, Span<byte> buffer, IEnumerable<KeyValuePair<string, string>> extendedAttributes, bool isGea)
+        {
+            // The ustar fields (uid, gid, linkName, uname, gname, devmajor, devminor) do not get written.
+            // The mode gets the default value.
+            _name = GenerateExtendedAttributeName();
+            _mode = (int)TarHelpers.DefaultMode;
+            _typeFlag = isGea ? TarEntryType.GlobalExtendedAttributes : TarEntryType.ExtendedAttributes;
+            _linkName = string.Empty;
+            _magic = string.Empty;
+            _version = string.Empty;
+            _gName = string.Empty;
+            _uName = string.Empty;
+
+            _dataStream = GenerateExtendedAttributesDataStream(extendedAttributes);
+
+            WriteAsPaxInternal(archiveStream, buffer);
+        }
+
+        // Both the Extended Attributes and Global Extended Attributes entry headers are written in a similar way, just the data changes
+        // This method writes an entry as both entries require, using the data from the current header instance.
+        private void WriteAsPaxInternal(Stream archiveStream, Span<byte> buffer)
+        {
+            long actualLength = GetTotalDataBytesToWrite();
+            TarEntryType actualEntryType = GetCorrectTypeFlagForFormat(TarFormat.Pax);
+
+            int checksum = WritePosixName(buffer);
+            checksum += WriteCommonFields(buffer, actualLength, actualEntryType);
+            checksum += WritePosixMagicAndVersion(buffer);
+            checksum += WritePosixAndGnuSharedFields(buffer);
+            WriteChecksum(checksum, buffer);
+
+            archiveStream.Write(buffer);
+
+            if (_dataStream != null)
+            {
+                WriteData(archiveStream, _dataStream, actualLength);
+            }
+        }
+
+        // All formats save in the name byte array only the ASCII bytes that fit. The full string is returned in the out byte array.
+        private int WriteName(Span<byte> buffer, out byte[] fullNameBytes)
```

Comment:
Multiple call sites ignore fullNameBytes.  Can we avoid allocating it at least in those cases?